extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1192,8 +1193,34 @@ private Int2ObjectMap<Object> makeWorkerFactoryInfosForStage(
   {
     if (MSQControllerTask.isIngestion(querySpec) &&
         stageNumber == queryDef.getFinalStageDefinition().getStageNumber()) {
-      // noinspection unchecked,rawtypes
-      return (Int2ObjectMap) makeSegmentGeneratorWorkerFactoryInfos(workerInputs, segmentsToGenerate);
+      final DataSourceMSQDestination destination = (DataSourceMSQDestination) querySpec.getDestination();
+      if (!destination.doesSegmentMorphing()) {
+        // noinspection unchecked,rawtypes
+        return (Int2ObjectMap) makeSegmentGeneratorWorkerFactoryInfos(workerInputs, segmentsToGenerate);
+      } else {
+        // worker info is the new lock version
+        if (destination.getReplaceTimeChunks().size() != 1) {
```

Comment:
Should this logic be present in the morphFactory itself since each morpher would require extra information according to the morpher needs ?

Reply:
Removed this part of the code.